bolt-modules/boltlib/lib/puppet/functions/background.rb

```diff
@@ -16,22 +16,21 @@
   #   }
   #   run_command("echo 'Continue immediately'", $targets)
   dispatch :background do
-    scope_param
     optional_param 'String[1]', :name
     block_param 'Callable[0, 0]', :block
     return_type 'Future'
   end
 
-  def background(scope, name = nil, &block)
+  def background(name = nil, &block)
     unless Puppet[:tasks]
       raise Puppet::ParseErrorWithIssue
         .from_issue_and_stack(Bolt::PAL::Issues::PLAN_OPERATION_NOT_SUPPORTED_WHEN_COMPILING, action: 'background')
     end
 
     executor = Puppet.lookup(:bolt_executor)
     executor.report_function_call(self.class.name)
-
-    executor.create_future(scope: scope, name: name) do |newscope|
+    plan_id = Puppet.lookup(:plan_stack).last
+    executor.create_future(scope: closure_scope, name: name, plan_id: plan_id) do |newscope|
```

Comment:
What is closure scope?

Reply:
It's a special variable that gets [injected into the function by Puppet when creating a custom function in Ruby](https://puppet.com/docs/puppet/6/functions_ruby_implementation.html#functions_ruby_implementation-accessing-puppet-vars) that represents the Puppet scope object. It's generally useful for accessing variables from the manifest or Bolt plan. Here we were trying to use it to give a unique ID to each Bolt plan invocation, since in theory the scope should be different for each invocation and it has been tricky to distinguish each plan *invocation* when the plan *name* can be the same (ie. calling the same plan multiple times). But this ended up not working because the scope can change based on any number of things (a new block, like `each {}`, etc.) and it was impossible to determine from the scope if function calls were in the same plan invocation or not.